fix a indices bug for categorical-datetime columns

alexifm opened the pull request:
This is to fix a bug reported in #26859

- [x] closes #26859
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

alexifm committed:
fix a indices bug for categorical-datetime columns

This is to fix a bug reported in #26859

WillAyd commented:
Can you add a test? That's typically the first thing we look for with PRs

alexifm committed:
test for DataFrameGroupby.indices

Adds a test for a bug fix for DataFrameGroupby.indices in  #26860

pep8speaks commented:
Hello @alexifm! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/groupby/test_groupby.py`](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py):

> [Line 1756:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1756): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1760:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1760): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1809:13](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1809): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1813:80](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1813): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1815:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1815): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1816:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1816): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1817:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1817): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1818:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1818): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1819:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1819): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1821:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1821): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1822:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1822): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1822:1](https://github.com/pandas-dev/pandas/blob/cfa0fefafbed98174eeddf45a3e3a2cc4f74f611/pandas/tests/groupby/test_groupby.py#L1822): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

alexifm committed:
pep8 formatting for test

cleans up the test to adhere to pep8 formatting

alexifm commented:
I need to find a better way to test to equality of the dictionaries in the output for 3.5.

alexifm committed:
Fix tests to handly Py3.5

The test no longer depends on ordering of a dictionary.  Also, the test matches the timestamp/datetime outputs that are the current standard in the code.

alexifm committed:
Fixing Py3.5 and other checks.

Handle Py3.5 dict ordering issues.  Cleanup for Pep8.  No longer using numpy testing utility.

jreback commented on pandas/core/groupby/ops.py:
```diff
@@ -232,6 +232,7 @@ def indices(self):
             label_list = [ping.labels for ping in self.groupings]
             keys = [com.values_from_object(ping.group_index)
                     for ping in self.groupings]
+            keys = [np.array(key) for key in keys]
```
this needs to be in get_indexer_dict in pandas/core/sorting.py

also, use ``np.asarray`` (np.array like this copies)

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1752,3 +1752,81 @@ def test_groupby_groups_in_BaseGrouper():
+    period_series = dt_series.to_period('Q')
+    period_series_cat = period_series.astype('category')
+
+    df = pd.DataFrame({
```
just do this inline (no need to name the Series)

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1752,3 +1752,81 @@ def test_groupby_groups_in_BaseGrouper():
     assert(result.groups == expected.groups)
+
+
+def test_groupby_indices():
```
move this after the test:  test_indices_concatenation_order()

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1752,3 +1752,81 @@ def test_groupby_groups_in_BaseGrouper():
+        'period_series': period_series,
+        'period_series_cat': period_series_cat
+    })
+    col_order = [
```
if you want to specify an ordering do it in the constructor itself

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1752,3 +1752,81 @@ def test_groupby_groups_in_BaseGrouper():
+    ]
+    n_choices = len(df.columns)
+
+    for n in range(1, n_choices + 1):
```
you can parameterize this function  rather than doing this as it makes the test very opaque

topper-123 commented:
This should also have a entry in the whatsnew.

alexifm commented:
Thanks for the feedback.  Will update.

jreback commented:
@alexifm can you merge master and update

alexifm committed:
updated groupby indices test to address comments

alexifm committed:
move test position

alexifm committed:
move bug fix to address comment

alexifm committed:
Merge branch 'master' into patch-1

alexifm committed:
undo pep8 autoformatting

alexifm committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

alexifm committed:
format fix

alexifm committed:
add bug to what's new

alexifm committed:
provide simple test for original github issue; simplify full output test; address py3.5 issues

alexifm committed:
update what's new for 1.0

alexifm committed:
address comments on cleaning up test

alexifm committed:
parametrize test

alexifm committed:
fix imports

alexifm committed:
import fix

alexifm committed:
black formatting

TomAugspurger commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -176,7 +176,7 @@ Groupby/resample/rolling
 
 -
--
+- Bug in :meth:`DataFrameGroupBy.indices` raises exception when grouping on multiple columns and one is a categorical with datetime values. (:issue:`26860`)
```
"raises" -> "raising an"

"and one" -> "when one"

TomAugspurger reviewed (commented):
Sorry, there's a merge conflict in the whatsnew. Can you merge master and resolve that?

@jbrockmendel do you have thoughts here?

jbrockmendel commented:
i think we need the test cases to include Categorical[datetimetz]

jreback commented:
can you merge master and update to comments

jreback commented:
can you merge master

jbrockmendel commented:
@alexifm can you rebase.  this would be nice to get in

jreback commented:
closing as stale, if you want to continue pls ping.
